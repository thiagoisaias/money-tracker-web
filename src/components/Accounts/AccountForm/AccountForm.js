import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import { devices } from "../../../utils/devices";
import Button from "../../Button/Button";
import Layout from "../../Layout/Layout";
import Spinner from "../../Spinner/Spinner";

const Container = styled.div`
  background-color: #fff;
  box-shadow: 1px;
  margin: 0 auto;
  color: #484848;
  box-shadow: 0.5px 1px 1px 1px #ddd;

  @media ${devices.small} {
    padding: 32px 16px;
  }

  @media ${devices.mediumUp} {
    padding: 32px 32px 55px 32px;
    margin: 32px auto;
    border-radius: 2px;
  }
`;

const Title = styled.h2`
  font-size: 24px;
  font-weight: 600;
  letter-spacing: 0.5px;
`;

const FormContainer = styled.form``;

const FormGroup = styled.div`
  display: flex;
  flex-direction: column;
  margin: 16px 0;

  @media ${devices.small} {
    width: 100%;
  }

  @media ${devices.mediumUp} {
    width: 45%;
    &:first-child {
      margin-right: 24px;
    }
  }
`;

const Label = styled.label`
  font-weight: 400;
  color: #777;
  margin-bottom: 6px;
`;

const Input = styled.input`
  border: 1px solid #f2f2f2;
  border-radius: 2px;
  padding: 6px;
  font: inherit;
  color: inherit;

  &:focus {
    border: 1px solid #ddd;
  }
`;

const Row = styled.div`
  @media ${devices.mediumUp} {
    display: flex;
    justify-content: space-between;
  }
`;

const ErrorMessage = styled.p`
  color: #e75252;
  font-size: 14px;
  margin-top: 16px;
`;

class AccountForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: this.props.name || "",
      initialBalance: this.props.initialBalance || ""
    };
  }

  handleInputChange = event => {
    const target = event.target;
    const value = target.value;
    const name = target.name;

    this.setState({
      [name]: value
    });
  };

  handleSubmit = event => {
    const { isLoading, submitData } = this.props;
    event.preventDefault();
    if (isLoading) {
      return;
    }
    submitData(this.state);
  };

  render() {
    const { error, isLoading, match } = this.props;
    return (
      <Layout>
        <Container>
          <Title>
            {match.path === "/accounts/new" ? "New Account" : "Edit Account"}
          </Title>
          {error && <ErrorMessage>{error}</ErrorMessage>}
          <FormContainer onSubmit={this.handleSubmit}>
            <Row>
              <FormGroup>
                <Label>{"Name"}</Label>
                <Input
                  type="text"
                  value={this.state.name}
                  name="name"
                  onChange={this.handleInputChange}
                />
              </FormGroup>
              <FormGroup>
                <Label>{"Initial Balance"}</Label>
                <Input
                  type="number"
                  value={this.state.initialBalance}
                  name="initialBalance"
                  onChange={this.handleInputChange}
                />
              </FormGroup>
            </Row>
            <Button onClick={this.handleSubmit}>
              {isLoading ? <Spinner height={34} width={34} /> : "Submit"}
            </Button>
          </FormContainer>
        </Container>
      </Layout>
    );
  }
}

AccountForm.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  error: PropTypes.array,
  initialBalance: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  match: PropTypes.object.isRequired,
  submitData: PropTypes.func.isRequired
};

export default AccountForm;
