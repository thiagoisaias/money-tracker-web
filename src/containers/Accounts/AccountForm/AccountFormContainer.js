import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import AccountForm from "../../../components/Accounts/AccountForm/AccountForm";

import { createAccount } from "../../../store/actions/accounts/accounts";

export class AccountFormContainer extends Component {
  submitData = formData => {
    const { userId, onCreateAccount } = this.props;
    
    onCreateAccount(formData, userId);
  };

  render() {
    const { error, isLoading } = this.props;
    return (
      <AccountForm
        submitData={this.submitData}
        error={error}
        isLoading={isLoading}
      />
    );
  }
}

AccountFormContainer.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  error: PropTypes.string,
  userId: PropTypes.number.isRequired,
  onCreateAccount: PropTypes.func.isRequired
};

const mapStateToProps = state => {
  return {
    userId: state.auth.user.id,
    isLoading: state.accounts.isLoading,
    error: state.accounts.error
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onCreateAccount: (accountData, userId) => {
      dispatch(createAccount(accountData, userId));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(
  AccountFormContainer
);
