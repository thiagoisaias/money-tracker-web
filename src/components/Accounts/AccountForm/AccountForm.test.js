import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { AccountForm } from "./AccountForm";

describe("AccountForm", () => {
  const mockProps = {
    userId: "UASDH19283",
    isLoading: false,
    error: null,
    onCreateAccount: jest.fn(),
    onClearAccountState: jest.fn()
  };
  const component = shallow(<AccountForm {...mockProps} />);

  it("renders properly", () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
