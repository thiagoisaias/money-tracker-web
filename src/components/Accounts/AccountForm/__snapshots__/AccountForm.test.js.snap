// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountForm renders properly 1`] = `
<AccountForm
  checkFormValidity={[Function]}
  error={null}
  handleInputChange={[Function]}
  id={1}
  initialBalance="32423.43"
  isLoading={false}
  match={
    Object {
      "path": "/path",
    }
  }
  name="Lorem"
  submitData={[Function]}
/>
`;
